enhancements/update/standalone-worker-surge.md

```diff
@@ -0,0 +1,295 @@
+and "drain" them (attempt to migrate workloads by rescheduling its pods onto uncordoned nodes). Workers generally need to be rebooted during a
+cluster update and draining nodes is best practice before rebooting them. If they were not drained first, pods running
+on a node targeted by the update process could be terminated with no other viable pods on the cluster to 
+handle the workload. This outcome can cause a disruption in the service the terminated pod was attempt to provide. For example,
+an incoming web request may not be routable to a pod for a given Kubernetes service - resulting in errors being returned
+to the consumers of that service.
+
+With appropriate cluster management, node draining can be used to ensure that sufficient pods are running to satisfy workload requirements
+at all times - even during updates. "Appropriate cluster management," though, is a multi-faceted challenge involving considerations
+from pod replicas to cluster topology.
+
+### Managing Worker Node Capacity
+
+One aspect of this challenge is ensuring that, while a node is being drained, there is sufficient worker node capacity (CPU/memory/other
+resources) available for new pods to take the place of old pods from the node being drained. Consider the reductive example of 
+a static cluster with a single worker node. If there is an attempt to drain the node in this example, there is no additional worker 
+node capacity available to schedule new pods to replace the pods being drained. This can result in a stalled drain -- one that 
+does not terminate until there is external intervention.
+
+Stalled drains create a frustrating experience for operations teams as they require analysis and intervention. They can also
+make it impossible to predict when an update will complete -- complicating work schedules and communications. There are a number of reasons 
+drains can stall, but simple lack of spare worker node capacity is a common one. The reason is that running an over-provisioned cluster
+(i.e. with more worker nodes than actually required for the steady state workloads) is not always cost-effective. Running
+an extra node 24x7 simply to ensure capacity for the short period while another node is being drained is inefficient.
+
+One cost-effective approach to ensure capacity is called "surging". With a surge strategy, during an update, the platform
+is permitted to bring additional worker nodes online, to accommodate workloads being drained from existing nodes. After an 
+update concludes, the surged nodes are scaled down and the cluster resumes its steady state.
+
+HyperShfit Hosted Control Planes (HCP) already support the surge concept. HyperShift `NodePools`
+expose `maxUnavailable` and `maxSurge` as configurable options during updates: https://hypershift-docs.netlify.app/reference/api/#hypershift.openshift.io/v1beta1.RollingUpdate . 
+Unfortunately, standalone OpenShift, which uses `MachineConfigPools`, does not. To workaround this

```

Comment:
I think it would probably be preferable all round to not duplicate the features here since we want MachineDeployment in the standalone product in the future. Perhaps instead of spending cycles implementing this here, the time could be spent helping to bring MachineDeployment in and solving the issues within OCP that this would cause (like the MCS being able to serve a particular new ignition version of the rendered configuration)

Reply:
Hypershift does indeed replace all workers. I think one thing we can look to do in standalone is allow a hybrid update model, where you can configure maxSurge and maxUnavailable on top.

What I mean by that is, let's say you have 6 worker nodes, and you want the end result to be 6 worker nodes, and you don't really care how that's achieved, so long as:
1. you never go above 9 workers (maxSurge of 3), and
2. 4 workers are always available for scheduling (maxUnavailable of 4)

In that case something in the cluster can configure it such that:
1. some nodes are updated inplace, in parallel
2. some nodes are surge'd in to the cluster on the newest configuration directly
3. some nodes are simply removed

In that case this something needs to preserve the maxSurge and maxUnavailable we set above. Could that be done via a hybrid of CAPI and MCO? Or should there be a central management point for this?